presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java

```diff
@@ -162,19 +164,26 @@ public void setOutputBuffers(OutputBuffers newOutputBuffers)
             if (outputBuffers.isNoMoreBufferIds()) {
                 this.state.compareAndSet(OPEN, NO_MORE_BUFFERS);
                 this.state.compareAndSet(NO_MORE_PAGES, FLUSHING);
+
+                pagesToDereference = ImmutableList.copyOf(initialPagesForNewBuffers);
```

Comment:
Note to self: this effectively moves `noMoreBuffers` to the same synchronized block.